### ISSUE TITLE
Analyzer should flag use of reference types in finalizer or Dispose(false)

### ISSUE THREAD

**Issue by AArnott** [DotNetAnalyzers/IDisposableAnalyzers#141]
Any use of reference types from a finalizer is hazardous. When that finalizer was invoked as part of AppDomain shutdown, the CLR makes *no guarantee* regarding order of finalization or GC for a typical finalizer, and thus any access to a reference type (other than strictly `this`) could AV and crash the process. The only safe activity from a finalizer is accessing value types and calling p/invoke methods, which limits safe activity to calling into native code to release resources. Even accessing SafeHandles is unsafe, which is why these types have their own finalizers rather than relying on their owners to dispose of them during finalization.

We should have an analyzer to flag any use of reference types in the finalizer or the `Dispose(bool)` method outside the `if (disposing)` block. For example, each of these methods represent a violation of the rule:

```cs
class MyType
{
  ILogger logger;

  ~MyType()
  {
    Dispose(false);
    this.logger.Log("Being finalized!"); // violation! ILogger is a reference type!  
  }

  protected virtual void Dispose(bool disposing)
  {
     if (disposing)
     {
     }

     this.logger.Log("In Dispose(bool)"); // violation! ILogger is a ref type and we're not inside the above block
  }
}
```

**Comment by JohanLarsson** [comment]
I'm using your text in the documentation, that makes it the first analyzer with good docs 😄
